README
incore_twitter is an extension for the TYPO3 Web Content Management System (http://www.typo3.org), providing a fancy, jQuery-based Twitter Ticker. The frontend look and feel of incore_twitter can be fully customized via CSS, while the extension itself can be configured via the TYPO3 backend or directly via TypoScript.
Development of this extension was heavily influenced by the jQuery Twitter Ticker presented and discussed on Tutorialzine (http://tutorialzine.com/2009/10/jquery-twitter-ticker/). See the original demo at http://demo.tutorialzine.com/2009/10/jquery-twitter-ticker/demo.html


ChangeLog
2011-10-28 incore IT Solutions <office@incore.at>
	* added CSS selectors for first and last tweet
	  (thanks to Tomas Norre Mikkelson) 

2011-10-27 incore IT Solutions <office@incore.at>
	* updated jScrollPane to 2.0.0beta10

2011-10-02 incore IT Solutions <office@incore.at>
	* added feature to limit search results to certain hashtags
	* updated documentation 

2010-07-27 incore IT Solutions <office@incore.at>
	* code cleanup (removed deprecated functions)

2010-06-02 incore IT Solutions  <office@incore.at>
	* changed status to 'stable'
	* updated description
	* updated documentation

2010-05-31 incore IT Solutions  <office@incore.at>
	* initial codebase uploaded to TER



ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY.'_pi1']='layout,select_key';


t3lib_extMgm::addPlugin(array(
	'LLL:EXT:incore_twitter/locallang_db.xml:tt_content.list_type_pi1',
	$_EXTKEY . '_pi1',
	t3lib_extMgm::extRelPath($_EXTKEY) . 'ext_icon.gif'
),'list_type');

// Use FlexForms
$TCA['tt_content']['types']['list']['subtypes_addlist'][$_EXTKEY . '_pi1'] = 'pi_flexform';
t3lib_extMgm::addPiFlexFormValue($_EXTKEY . '_pi1', 'FILE:EXT:' . $_EXTKEY . '/pi1/flexform.xml');

// static CSS  
t3lib_extMgm::addStaticFile($_EXTKEY,"pi1/static/css","Twitter Ticker CSS");

if (TYPO3_MODE == 'BE') {
	$TBE_MODULES_EXT['xMOD_db_new_content_el']['addElClasses']['tx_incoretwitter_pi1_wizicon'] = t3lib_extMgm::extPath($_EXTKEY).'pi1/class.tx_incoretwitter_pi1_wizicon.php';
}
?>

js/jQueryTwitter.js
function TweetTick(ob)
{
	//var container=$('#tweet-container');
	//console.log(containerId);
	var container=$(containerId);
	var lastidx = 0;
	container.html('');
	if (ob.results.length != ob.results_per_page) {
		lastidx = ob.results.length-1;
	}
	else {
		lastidx = ob.results_per_page-1;
	}  	
	$(ob.results).each(function(index,el){
		var css = '';
        if(index == 0){
            css = ' first';
        }
        if(index == lastidx){
            css = ' last';
        }
        var str = ' <div class="tweet '+css+'">\
					<div class="avatar"><a href="http://twitter.com/'+this.from_user+'" target="_blank"><img src="'+this.profile_image_url+'" alt="'+this.from_user+'" /></a></div>\
					<div class="user"><a href="http://twitter.com/'+this.from_user+'" target="_blank">'+this.from_user+'</a></div>\
					<div class="time">'+relativeTime(this.created_at)+'</div>\
					<div class="txt">'+formatTwitString(this.text)+'</div>\
					</div>';
		container.append(str);
	});
	if (jQuery().jScrollPane){
		container.jScrollPane();
	}
}

function formatTwitString(str)
{
	str=' '+str;
	str = str.replace(/((ftp|https?):\/\/([-\w\.]+)+(:\d+)?(\/([\w/_\.]*(\?\S+)?)?)?)/gm,'<a href="$1" target="_blank">$1</a>');
	str = str.replace(/([^\w])\@([\w\-]+)/gm,'$1@<a href="http://twitter.com/$2" target="_blank">$2</a>');
	str = str.replace(/([^\w])\#([\w\-]+)/gm,'$1<a href="http://twitter.com/search?q=%23$2" target="_blank">#$2</a>');
	return str;
}

function relativeTime(pastTime)
{	
	var origStamp = Date.parse(pastTime);
	var curDate = new Date();
	var currentStamp = curDate.getTime();
	
	var difference = parseInt((currentStamp - origStamp)/1000);

	if(difference < 0) return false;

	if(difference <= 5)				return "Just now";
	if(difference <= 20)			return "Seconds ago";
	if(difference <= 60)			return "A minute ago";
	if(difference < 3600)			return parseInt(difference/60)+" minutes ago";
	if(difference <= 1.5*3600) 		return "One hour ago";
	if(difference < 23.5*3600)		return Math.round(difference/3600)+" hours ago";
	if(difference < 1.5*24*3600)	return "One day ago";
	
	var dateArr = pastTime.split(' ');
	return dateArr[4].replace(/\:\d+$/,'')+' '+dateArr[2]+' '+dateArr[1]+(dateArr[3]!=curDate.getFullYear()?' '+dateArr[3]:'');
}

pi1/class.tx_incoretwitter_pi1.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2011 incore IT Solutions <office@incore.at>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */

require_once(PATH_tslib.'class.tslib_pibase.php');


/**
 * Plugin 'jQuery Twitter Feed' for the 'incore_twitter' extension.
 *
 * @author	incore IT Solutions <office@incore.at>
 * @package	TYPO3
 * @subpackage	tx_incoretwitter
 */
class tx_incoretwitter_pi1 extends tslib_pibase {
	var $prefixId      = 'tx_incoretwitter_pi1';		// Same as class name
	var $scriptRelPath = 'pi1/class.tx_incoretwitter_pi1.php';	// Path to this script relative to the extension dir.
	var $extKey        = 'incore_twitter';	// The extension key.
	var $pi_checkCHash = true;

	/**
	 * The main method of the PlugIn
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	function main($content, $conf) {
		$this->init();
		$this->conf = $conf;
		$this->pi_setPiVarDefaults();
		$this->pi_loadLL();

		$this->lConf;

		if ($GLOBALS['TSFE']->additionalHeaderData[$this->prefixId] != '') {
			return '<strong>This Plugin can only be used once per page!</strong>';
		}
		$GLOBALS['TSFE']->additionalHeaderData[$this->prefixId] = '
			<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.4/jquery.min.js"></script>
			<script type="text/javascript" src="typo3conf/ext/incore_twitter/js/jScrollPane/jquery.mousewheel.js"></script>
			<script type="text/javascript" src="typo3conf/ext/incore_twitter/js/jScrollPane/jquery.jscrollpane.min.js"></script>		
			<script type="text/javascript" src="typo3conf/ext/incore_twitter/js/jQueryTwitter.js"></script>
		';
		$tweetUsersArray = t3lib_div::trimExplode(',',$this->lConf['twitteraccounts']);
		$hashtagsArray = t3lib_div::trimExplode(',', $this->lConf['hashtags']);
		$tweetUsers = "['" . implode("', '", $tweetUsersArray) . "']";
		$hashtags = "['" . implode("', '", $hashtagsArray) . "']";
		$uid = $this->cObj->data['uid'];
		$tweetLimit = $this->lConf['twittercounts'];
		if ($tweetLimit == 0) { $tweetLimit = 10;}
		
		$content = '
			<script type="text/javascript">
			var tweetUsers = ' . $tweetUsers . ';
			var hashTags = ' . $hashtags . ';
			var tweetsCount = ' . $tweetLimit . ';
			var buildString = "";
			var hashTagString = "";
			var URI = "";
			var containerId = \'#tweet-container-' . $uid . '\';
			$(document).ready(function(){	
				for(var i=0;i<tweetUsers.length;i++) {
					if(i!=0) buildString+=\'+OR+\';
					buildString+=\'from:\'+tweetUsers[i];
				}
				for(var j=0;j<hashTags.length;j++) {
					if(j!=0) hashTagString+=\'+OR+\';
					hashTagString+=\'%23\'+hashTags[j];
				}
				URI = "http://search.twitter.com/search.json?q=" + hashTagString + "+" + buildString + "&callback=TweetTick&rpp=" +tweetsCount;
				var fileref = document.createElement(\'script\');
				fileref.setAttribute("type","text/javascript");
				fileref.setAttribute("src", URI);
				document.getElementsByTagName("head")[0].appendChild(fileref);
			});
			</script>
			<div id="tweet-container-'.$uid.'" class="tweet-container">Loading..</div>
			<div id="scroll"></div>
			';
		return $this->pi_wrapInBaseClass($content);
	}

	function init(){
		$this->pi_initPIflexForm(); // Init and get the flexform data of the plugin
		$this->lConf = array(); // Setup our storage array...
		// Assign the flexform data to a local variable for easier access
		$piFlexForm = $this->cObj->data['pi_flexform'];
		// Traverse the entire array based on the language...
		// and assign each configuration option to $this->lConf array...
		foreach ($piFlexForm['data'] as $sheet => $data ) {
			foreach ( $data as $lang => $value ) {
				foreach ( $value as $key => $val ) {
					$this->lConf[$key] = $this->pi_getFFvalue($piFlexForm, $key, $sheet);
				}
			}
		}
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/incore_twitter/pi1/class.tx_incoretwitter_pi1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/incore_twitter/pi1/class.tx_incoretwitter_pi1.php']);
}

?>


pi1/class.tx_incoretwitter_pi1_wizicon.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 incore IT Solutions <office@incore.at>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */




/**
 * Class that adds the wizard icon.
 *
 * @author	incore IT Solutions <office@incore.at>
 * @package	TYPO3
 * @subpackage	tx_incoretwitter
 */
class tx_incoretwitter_pi1_wizicon {

					/**
					 * Processing the wizard items array
					 *
					 * @param	array		$wizardItems: The wizard items
					 * @return	Modified array with wizard items
					 */
					function proc($wizardItems)	{
						global $LANG;

						$LL = $this->includeLocalLang();

						$wizardItems['plugins_tx_incoretwitter_pi1'] = array(
							'icon'=>t3lib_extMgm::extRelPath('incore_twitter').'pi1/ce_wiz.gif',
							'title'=>$LANG->getLLL('pi1_title',$LL),
							'description'=>$LANG->getLLL('pi1_plus_wiz_description',$LL),
							'params'=>'&defVals[tt_content][CType]=list&defVals[tt_content][list_type]=incore_twitter_pi1'
						);

						return $wizardItems;
					}

					/**
					 * Reads the [extDir]/locallang.xml and returns the $LOCAL_LANG array found in that file.
					 *
					 * @return	The array with language labels
					 */
					function includeLocalLang()	{
						$llFile = t3lib_extMgm::extPath('incore_twitter').'locallang.xml';
						$LOCAL_LANG = t3lib_div::readLLXMLfile($llFile, $GLOBALS['LANG']->lang);

						return $LOCAL_LANG;
					}
				}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/incore_twitter/pi1/class.tx_incoretwitter_pi1_wizicon.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/incore_twitter/pi1/class.tx_incoretwitter_pi1_wizicon.php']);
}

?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "incore_twitter".
#
# Auto generated 11-01-2012 07:15
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'jQuery Twitter Ticker',
	'description' => 'This extension provides a fancy jQuery-based Twitter Ticker, featuring the latest Tweets of one or more Twitter accounts, as a Frontend Plugin.',
	'category' => 'plugin',
	'shy' => 0,
	'version' => '1.2.2',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'incore IT Solutions',
	'author_email' => 'office@incore.at',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:24:{s:9:"ChangeLog";s:4:"cb60";s:10:"README.txt";s:4:"33e2";s:12:"ext_icon.gif";s:4:"24bd";s:17:"ext_localconf.php";s:4:"8fd2";s:14:"ext_tables.php";s:4:"eb55";s:14:"ext_tables.sql";s:4:"d41d";s:30:"icon_tx_incoretwitter_data.gif";s:4:"d41d";s:13:"locallang.xml";s:4:"038e";s:16:"locallang_db.xml";s:4:"1ace";s:7:"tca.php";s:4:"d41d";s:14:"doc/manual.sxw";s:4:"cb22";s:19:"doc/wizard_form.dat";s:4:"fdce";s:20:"doc/wizard_form.html";s:4:"e5f7";s:19:"js/jQueryTwitter.js";s:4:"8b88";s:30:"js/jScrollPane/MIT-LICENSE.txt";s:4:"adaf";s:40:"js/jScrollPane/jquery.jscrollpane.min.js";s:4:"5b97";s:35:"js/jScrollPane/jquery.mousewheel.js";s:4:"ec36";s:14:"pi1/ce_wiz.gif";s:4:"a3ed";s:34:"pi1/class.tx_incoretwitter_pi1.php";s:4:"718d";s:42:"pi1/class.tx_incoretwitter_pi1_wizicon.php";s:4:"7b6f";s:13:"pi1/clear.gif";s:4:"cc11";s:16:"pi1/flexform.xml";s:4:"fe6a";s:17:"pi1/locallang.xml";s:4:"527c";s:24:"pi1/static/css/setup.txt";s:4:"6673";}',
);

?>